scripts/cangen/cangen/file_parser.py

```diff
@@ -0,0 +1,118 @@
+from __future__ import annotations
+import json
+import yaml
+from enum import Enum
+from typing import Dict, Any
+from database import (
+    Bus,
+    Node,
+    Message,
+    Signal,
+    IntegralSignal,
+    BooleanSignal,
+    EnumSignal,
+    FloatingSignal,
+)
+
+
+def _parse_signal(signal_data: Dict[str, Any], enum_types: Dict[str, Enum]) -> Signal:
+    signal_type = signal_data.get("signal_type")
+
+    # Create appropriate signal class based on signal type
+    signal_classes = {
+        "integral": IntegralSignal,
+        "boolean": BooleanSignal,
+        "enum": EnumSignal,
+        "floating": FloatingSignal,
+    }
+
+    # Create the signal instance with appropriate context for validation
+    parse_context = None
+    if signal_type == "enum":
+        parse_context = {"enum_types": enum_types}
+
+    return signal_classes[signal_type](signal_data, context=parse_context)
+
+
+def _parse_message(
+    message_data: Dict[str, Any], nodes: Dict[str, Node], enum_types: Dict[str, Enum]
+) -> Message:
+    # Parse and create a Message object with its signals and node references.
+    signal_data_list = message_data.get("signals", [])
+    message_data_copy = {k: v for k, v in message_data.items() if k != "signals"}
+
+    message = Message(message_data_copy, context={"nodes": nodes.keys()})
```

Comment:
```suggestion
    signal_data_list = message_data.pop("signals")

    message = Message(message_data, context={"nodes": nodes.keys()})
```

two things here:
1. the `signals` field should be required on each message. if the user wants it empty, they can explicitly write `signals: []` in the yaml
2. Use `.pop` to remove the element from the dictionary. then you don't need to make a copy (just continue using `message_data`). In general, you should always `pop` the data to consume it instead of temporarily `get`-ting it